create_derivatives/routines.py

```diff
@@ -69,9 +70,104 @@ def get_as_uri(self, bag_filepath):
 
 
 class JP2Maker:
-    # TO DO: make JPG2000 derivatives
+    """Creates JP2000 derivatives from TIFFs.
+
+    Creates JP2 directory in bag's data directory and JP2 derivatives.
+    TIFFs to be converted are targeted in the service directory. If the service directory is empty or
+    does not exist, target TIFFs at the root of the data directory.
+
+    Returns:
+        A tuple containing human-readable message along with list of bag identifiers.
+        Exceptions are raised for errors along the way.
+    """
+
     def run(self):
-        pass
+        bags_with_jp2s = []
+        for bag in Bag.objects.filter(process_status=Bag.PREPARED):
+            jp2_dir = Path(bag.bag_path, "data", "JP2")
+            if not jp2_dir.is_dir():
+                jp2_dir.mkdir()
+            service_dir = Path(bag.bag_path, "data", "service")
+            if service_dir.is_dir() and any(service_dir.iterdir()):
+                tiff_files_dir = Path(bag.bag_path, "data", "service")
+            else:
+                tiff_files_dir = Path(bag.bag_path, "data")
+            self.create_jp2(bag, tiff_files_dir, jp2_dir)
+            bag.process_status = Bag.JPG2000
+            bag.save()
+            bags_with_jp2s.append(bag.bag_identifier)
+        msg = "JPG2000s created." if len(bags_with_jp2s) else "No TIFF files ready for JP2 creation."
+        return msg, bags_with_jp2s
+
+    def calculate_layers(self, file):
+        """Calculates the number of layers based on pixel dimensions.
+        For TIFF files, image tag 256 is the width, and 257 is the height.
+
+        Args:
+            file (str): filename of a TIFF image file.
+        Returns:
+            layers (int): number of layers to convert to
+        """
+
+        with Image.open(file) as img:
+            width = [w for w in img.tag[256]][0]
+            height = [h for h in img.tag[257]][0]
+        return math.ceil((math.log(max(width, height)) / math.log(2)
+                          ) - ((math.log(96) / math.log(2)))) + 1
+
+    def get_page_number(self, file):
+        """Parses a page number from a filename.
+
+        Presumes that:
+            The page number is preceded by an underscore
+            The page number is immediately followed by either by `_m`, `_me` or `_se`,
+            or the file extension.
+
+        Args:
+            file (str): filename of a TIFF image file.
+        Returns:
+            page number from the filename
+        """
+        filename = Path(file).stem
+        if "_se" in filename:
+            filename_trimmed = filename.split("_se")[0]
+        elif "_m" in filename:
+            filename_trimmed = filename.split("_m")[0]
+        else:
+            filename_trimmed = filename
+        return filename_trimmed.split("_")[-1]
+
+    def create_jp2(self, bag, tiff_files_dir, jp2_dir):
```

Comment:
Very minor nitpick but should this method be `create_jp2s` since it's producing multiple files?